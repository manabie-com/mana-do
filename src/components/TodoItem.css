.ToDo__item {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-top: 15px;
    padding-bottom: 15px;
    border-bottom: 1px rgb(141, 140, 140) solid;
}

.ToDo__item > span {
    flex: 1 1;
    text-align: left;
    margin-left: 8px;
}
.Todo__delete {
    outline: none;
    border: none;
    width: 32px;
    height: 32px;
    min-width: auto;
    min-height: auto;
    box-shadow: none;
    border-radius: 50%;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
}
