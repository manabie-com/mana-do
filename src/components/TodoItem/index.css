.ToDo__list {
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;
}

.ToDo__item {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 10px;
  border-radius: 7px;
  padding: 2px 5px 2px 5px;
}

.ToDo__item input {
  background-color: transparent;
  box-shadow: none;
}

.ToDo__item > span {
  flex: 1 1;
  text-align: left;
  margin-left: 8px;
}
