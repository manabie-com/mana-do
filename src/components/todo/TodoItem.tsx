import React, { useEffect, useRef, useState } from 'react';
import { TodoStatus, Todo } from '../../models/todo';
import useClickOutside from '../../hooks/useClickOutside';

interface TodoItemProps {
  todo: Todo;
  onDelete?: (id: string) => void;
  onComplete?: (id: string, isComplete: boolean) => void;
  onUpdateContent?: (id: string, content: string) => void;
}

const TodoItem = (props: TodoItemProps) => {
  const [isUpdateContent, setIsUpdateContent] = useState<boolean>(false);
  const [editingContent, setEditingContent] = useState<string>();
  const itemRef = useRef<HTMLDivElement>(null);
  const editingContentRef = useRef<HTMLTextAreaElement>(null);

  useClickOutside(itemRef, () => {
    if (isUpdateContent === true) {
      setIsUpdateContent(false);
    }
  });

  useEffect(() => {
    if (isUpdateContent) {
      const length = editingContentRef.current?.value.length || null;
      editingContentRef.current?.setSelectionRange(length, length);
      editingContentRef.current?.focus();
    }
  }, [isUpdateContent]);

  const onEnableUpdateContent = (e: React.MouseEvent<HTMLParagraphElement>) => {
    // double click
    if (e.detail === 2) {
      setEditingContent(props.todo.content);
      setIsUpdateContent(true);
    }
  };

  const onUpdateTodoContent = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    if (e.key === 'Enter' && editingContent?.trim()) {
      props.onUpdateContent &&
        props.onUpdateContent(props.todo.id, editingContent.trim());
      setIsUpdateContent(false);
    }
  };

  return (
    <div ref={itemRef} className="item">
      {isUpdateContent ? (
        <textarea
          ref={editingContentRef}
          className="item__content item__content-edit"
          value={editingContent}
          onChange={(e) => setEditingContent(e.target.value)}
          onKeyDown={onUpdateTodoContent}
        />
      ) : (
        <p
          className="item__content item__content-noedit"
          onClick={onEnableUpdateContent}
        >
          {props.todo.content}
        </p>
      )}

      <div className="item__actions">
        <button
          className="item__delete"
          onClick={() => props.onDelete && props.onDelete(props.todo.id)}
        >
          X
        </button>
        <div className="action__checkbox">
          <input
            id={`completed-${props.todo.id}`}
            type="checkbox"
            checked={TodoStatus.COMPLETED === props.todo.status}
            onChange={(e) =>
              props.onComplete &&
              props.onComplete(props.todo.id, e.target.checked)
            }
          />
          <label htmlFor={`completed-${props.todo.id}`}>Completed</label>
        </div>
      </div>
    </div>
  );
};

export default TodoItem;
