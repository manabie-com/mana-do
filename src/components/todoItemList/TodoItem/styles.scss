.todo-item {
  display: flex;
  align-items: center;
  justify-content: space-between;

  & > span {
    flex: 1 1;
    text-align: left;
    margin-left: 8px;
  }

  &__input-wrapper {
    display: flex;
    align-items: center;

    & > label {
      margin-left: 10px;
    }
  }

  &__delete {
    outline: none;
    border: none;
    width: 32px;
    height: 32px;
    min-width: auto;
    min-height: auto;
    box-shadow: none;
    border-radius: 50%;
    display: flex;
    background-color: #fff;
    justify-content: center;
    align-items: center;
    cursor: pointer;

    &:hover {
      background-color: transparent;
    }
  }

  &__text {
    &--completed {
      text-decoration: line-through;
      color: #64748b;
    }
  }

  &__input {
    width: 100%;
  }
}
