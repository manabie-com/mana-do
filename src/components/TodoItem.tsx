import React, {Dispatch, useRef, useState} from "react";
import {
    AppActions,
    deleteTodo,
    updateTodoContent,
    updateTodoStatus
} from "store/actions";
import {Todo} from "models/todo";
import {isTodoCompleted} from "utils";

interface ToolBarProps {
    dispatch: Dispatch<AppActions>,
    todo: Todo,
}

const TodoItem: React.FC<ToolBarProps> = ({dispatch, todo}) => {
    const [editing, setEditing] = useState({id: '', text: ''});
    const editTodoRef = useRef<any>(null);

    const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
        dispatch(updateTodoStatus(todoId, e.target.checked))
    }

    const onDeleteTodo = (todoId: string) => {
        dispatch(deleteTodo(todoId))
    }

    const onToggleEditTodo = (todo: Todo) => {
        if (isTodoCompleted(todo)) return;
        setEditing({
            id: todo.id,
            text: todo.content
        })
    }

    const onUpdateTodoContentWhenEnter = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key !== 'Enter') return;
        dispatch(updateTodoContent(editing.id, editing.text))
        setEditing({id: '', text: ''})
    }
    const onDiscardEditTodo = (e: React.FocusEvent<HTMLInputElement>) => {
        setEditing({id: '', text: ''})
    }

    const editable = editing?.id && editing?.id === todo.id;
    const todoItemClassName = ['Todo__item', isTodoCompleted(todo) ? '__complete' : '', editable ? '__editing' : ''].join(' ');
    return (
        <div className={todoItemClassName}>
            {!editable && <input
                className="Todo__item__toggle"
                type="checkbox"
                checked={isTodoCompleted(todo)}
                onChange={e => onUpdateTodoStatus(e, todo.id)}
            />}
            <div className="Todo__item__content">
                <div className="__text" onDoubleClick={() => onToggleEditTodo(todo)}>{todo.content}</div>
            </div>
            {!editable &&
            <button title="Delete" className="Todo__item__delete" onClick={() => onDeleteTodo(todo.id)}>Ã—</button>}
            {editable && <input
                ref={editTodoRef}
                className="__input_edit"
                type="text"
                autoFocus
                defaultValue={editing?.text}
                onChange={e => setEditing({id: todo.id, text: e.target.value})}
                onBlur={e => onDiscardEditTodo(e)}
                onKeyDown={e => onUpdateTodoContentWhenEnter(e)}
            />}
        </div>
    );
};

export default TodoItem;
