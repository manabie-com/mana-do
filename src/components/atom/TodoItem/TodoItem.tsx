import React from "react";
import Checkbox from "components/atom/Checkbox";
import { Todo } from "models/todo";
import { deleteTodo } from "actions/TodoListAction";
import { isTodoCompleted } from "utils";
import GarbageSVG from "svgs/Garbage";
import EditSVG from "svgs/Edit";
import useTodoItem from "./hook";

type ToDoItemProps = {
  todo: Todo;
};

const ToDoItem: React.FC<ToDoItemProps> = ({ todo }) => {
  const {
    updateContent,
    onUpdateToDoByKeydown,
    onUpdateTodoStatus,
    setUpdateContent,
    dispatch,
    inputRef,
  } = useTodoItem(todo);

  return (
    <React.Fragment>
      {updateContent ? (
        <div key={todo.id} className="ToDo__item ToDo__item--edit">
          <input
            ref={inputRef}
            className="Todo__input"
            data-testid="todo_input"
            defaultValue={todo.content}
            onKeyPress={onUpdateToDoByKeydown}
            autoFocus
          />
        </div>
      ) : (
        <div key={todo.id} className="ToDo__item">
          <Checkbox
            checked={isTodoCompleted(todo)}
            onChange={(e) => onUpdateTodoStatus(e, todo.id)}
            label={todo.content}
            name={todo.id}
            id={todo.id}
            fullWidth={true}
          />
          <div className="Todo__actions">
            <button
              className="Todo__action Todo__action--edit"
              onClick={() => setUpdateContent(true)}
            >
              <EditSVG />
            </button>
            <button
              className="Todo__action Todo__action--delete"
              onClick={() => dispatch(deleteTodo(todo.id))}
            >
              <GarbageSVG />
            </button>
          </div>
        </div>
      )}
    </React.Fragment>
  );
};

export default ToDoItem;
