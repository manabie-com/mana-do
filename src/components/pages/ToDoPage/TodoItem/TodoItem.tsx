import React, { useState } from 'react';
import { Todo } from '../../../../models/todo';
import { AppActions, deleteTodo, editTodo, updateTodoStatus } from '../../../../store/actions';
import { isTodoCompleted } from '../../../../utils';
import Button from '../../../atom/Button/Button';
import FormInput from '../../../atom/FormInput/FormInput';

import './todoItem.scss'

interface ITodo {
    todo: Todo
    dispatch: React.Dispatch<AppActions>
    todosLength: number
}

export const BackgroundColor = {
    isCompleted: 'rgb(0, 255, 204)',
    isActive: 'rgb(255, 147, 147)'
}

const TodoItem = ({ todo, dispatch, todosLength }: ITodo) => {
    const [content, setContent] = useState<string>(todo.content)
    const [readOnly, setReadOnly] = useState<boolean>(true)

    const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
        dispatch(updateTodoStatus(todoId, e.target.checked))
    }

    const handleEdit = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === 'Enter') {
            dispatch(editTodo(todo.id, content))
            setReadOnly(true)
        }
    }

    const handleDelete = () => {
        if (todosLength === 1) {
            localStorage.removeItem('task')
        }
        dispatch(deleteTodo(todo.id));
    }

    return (
        <div data-testid={todo.id} className="ToDo__item">
            <FormInput className='ToDo__item-checkbox' type="checkbox" onChange={(e) => onUpdateTodoStatus(e, todo.id)} checked={isTodoCompleted(todo)} />
            <FormInput
                //to set background base on status of task
                style={isTodoCompleted(todo) ?
                    { backgroundColor: BackgroundColor.isCompleted }
                    : { backgroundColor: BackgroundColor.isActive }}
                className='ToDo__item-value'
                readOnly={readOnly}
                value={readOnly ? todo.content : content}
                onChange={(e: any) => setContent(e.target.value)}
                onDbClick={() => setReadOnly(false)}
                onKeyDown={handleEdit}
                onBlur={() => {
                    setContent(todo.content)
                    setReadOnly(true)
                }}
            />
            <Button
                label='X'
                className="ToDo__item-delete"
                onClick={handleDelete} />
        </div>
    )
}

export default TodoItem;