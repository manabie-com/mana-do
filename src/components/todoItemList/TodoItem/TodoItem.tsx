import { Button } from "components/button";
import { Checkbox } from "components/checkbox";
import { Input } from "components/input";
import { useOnClickOutside } from "hooks";
import { InputType, Todo } from "models";
import React, { Dispatch, FC, memo, useRef, useState } from "react";
import { IoTrashBin } from "react-icons/io5";
import {
  deleteTodo,
  updateTodoContent,
  updateTodoStatus,
} from "store/action-handlers";
import { AppActions } from "store/actions";
import { getTodoStatus, isTodoCompleted } from "utils";
import "./styles.scss";

export interface ITodoItemProps extends Todo {
  dispatch: Dispatch<AppActions>;
  className?: string;
}

const TodoItem: FC<ITodoItemProps> = ({ className, dispatch, ...props }) => {
  const { id, content, status } = props;

  const inputRef = useRef<HTMLInputElement>(null);
  const [isEdit, setIsEdit] = useState<boolean>(false);

  /** Event handlers --- start */

  const handleUpdateTodoStatus = (
    e: React.ChangeEvent<HTMLInputElement>
  ): void => {
    dispatch(updateTodoStatus(id, getTodoStatus(e.target.checked)));
  };

  const handleDeleteTodoItem = (): void => {
    dispatch(deleteTodo(id));
  };

  const handleDoubleClickToEdit = () => {
    setIsEdit(true);
  };

  const handleUpdateTodo = async (
    e: React.KeyboardEvent<HTMLInputElement>
  ): Promise<void> => {
    if (e.key === "Enter" && inputRef.current) {
      const { value } = inputRef.current;
      if (!value) return;
      dispatch(updateTodoContent(id, value));
      setIsEdit(false);
    }
  };

  const handleClickOutsideInput = () => {
    setIsEdit(false);
  };

  useOnClickOutside(inputRef, handleClickOutsideInput);

  /** Event handlers --- end */

  return (
    <div
      className={`todo-item ${className ? className : ""}`}
      data-testid="todo-item"
      {...props}
    >
      {isEdit ? (
        <Input
          name={id}
          type={InputType.TEXT}
          ref={inputRef}
          className="todo-item__input todo-item__input--editing"
          defaultValue={content}
          onKeyDown={handleUpdateTodo}
        />
      ) : (
        <>
          <div className="todo-item__input-wrapper">
            <Checkbox
              id={id}
              checked={isTodoCompleted(status)}
              onChange={handleUpdateTodoStatus}
            />
            <label
              className={`todo-item__text ${
                isTodoCompleted(status) ? "todo-item__text--completed" : ""
              }`}
              htmlFor={id}
              data-testid="label-content"
              onDoubleClick={handleDoubleClickToEdit}
            >
              {content}
            </label>
          </div>
          <Button
            aria-label="Close"
            onClick={handleDeleteTodoItem}
            className="todo-item__delete"
            data-testid="todo-item-delete"
          >
            <IoTrashBin fontSize={24} color="#F22257" />
          </Button>
        </>
      )}
    </div>
  );
};

export default memo(TodoItem);
