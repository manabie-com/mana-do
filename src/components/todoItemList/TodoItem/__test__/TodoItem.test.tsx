import { fireEvent, render, RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TodoStatus } from "models";
import React from "react";
import { deleteTodo, updateTodoStatus } from "store/action-handlers";
import { getElementByTestId } from "utils/testUtils";
import TodoItem, { ITodoItemProps } from "../TodoItem";

const onDispatch = jest.fn();

const props = {
  dispatch: onDispatch,
  id: "id 1",
  user_id: "user_id 1",
  content: "content testing 1",
  status: TodoStatus.COMPLETED,
  created_date: "2022-05-08T04:44:10.054Z",
};

const renderComponent = ({ ...props }: ITodoItemProps): RenderResult => {
  return render(<TodoItem {...props} />);
};

describe("<TodoItem /> rendering", () => {
  it("should match snapshot", () => {
    const wrapper = renderComponent(props);
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render correcly", () => {
    renderComponent(props);
    const element = getElementByTestId("todo-item");
    expect(element).toBeInTheDocument();
  });

  it("should render a input with the default class", () => {
    renderComponent(props);
    const element = getElementByTestId("todo-item");
    expect(element).toHaveClass("todo-item");
  });
});

describe("<TodoItem /> interacting", () => {
  it("should handle onClick to trigger deleteTodo", () => {
    renderComponent(props);

    const element = getElementByTestId("todo-item-delete");
    userEvent.click(element);
    expect(props.dispatch).toHaveBeenCalledWith(deleteTodo(props.id));
  });

  it("should handle onClick checkbox in order to updateTodoStatus", () => {
    renderComponent(props);

    const checkboxElement = getElementByTestId("checkbox") as HTMLInputElement;
    fireEvent.click(checkboxElement, { target: { checked: true } });
    expect(props.dispatch).toHaveBeenCalledWith(
      updateTodoStatus(props.id, TodoStatus.ACTIVE)
    );
  });

  it("should handle onDoubleClick in order to make input becomes editable", () => {
    renderComponent(props);

    const element = getElementByTestId("label-content");
    const deleteElement = getElementByTestId("todo-item-delete");
    const checkboxElement = getElementByTestId("checkbox");
    const labelElement = getElementByTestId("label-content");

    fireEvent.doubleClick(element);

    const inputElement = getElementByTestId("input") as HTMLInputElement;

    expect(inputElement).toBeInTheDocument();
    expect(deleteElement).not.toBeInTheDocument();
    expect(labelElement).not.toBeInTheDocument();
    expect(checkboxElement).not.toBeInTheDocument();
    expect(inputElement.value).toBe(props.content);
    expect(inputElement).toHaveClass("todo-item__input--editing");
  });
});
