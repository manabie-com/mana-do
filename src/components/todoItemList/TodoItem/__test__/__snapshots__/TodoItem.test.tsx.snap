// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> rendering should match snapshot 1`] = `
<div>
  <div
    class="todo-item "
    content="content testing 1"
    created_date="2022-05-08T04:44:10.054Z"
    data-testid="todo-item"
    id="id 1"
    status="COMPLETED"
    user_id="user_id 1"
  >
    <div
      class="todo-item__input-wrapper"
    >
      <input
        checked=""
        class="cbx-todo "
        data-testid="checkbox"
        id="id 1"
        type="checkbox"
      />
      <label
        class="todo-item__text todo-item__text--completed"
        data-testid="label-content"
        for="id 1"
      >
        content testing 1
      </label>
    </div>
    <button
      aria-label="Close"
      class="btn-todo todo-item__delete"
      data-testid="todo-item-delete"
    >
      <svg
        color="#F22257"
        fill="currentColor"
        font-size="24"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(242, 34, 87);"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="80"
          rx="32"
          ry="32"
          width="448"
          x="32"
          y="48"
        />
        <path
          d="M74.45 160a8 8 0 00-8 8.83l26.31 252.56a1.5 1.5 0 000 .22A48 48 0 00140.45 464h231.09a48 48 0 0047.67-42.39v-.21l26.27-252.57a8 8 0 00-8-8.83zm248.86 180.69a16 16 0 11-22.63 22.62L256 318.63l-44.69 44.68a16 16 0 01-22.63-22.62L233.37 296l-44.69-44.69a16 16 0 0122.63-22.62L256 273.37l44.68-44.68a16 16 0 0122.63 22.62L278.62 296z"
        />
      </svg>
    </button>
  </div>
</div>
`;
